modules/pns/pns.c

```diff
@@ -0,0 +1,146 @@
+/**
+ * @file pns.c Push Notification Service Module for SIP
+ *
+ * Copyright (C) 2023 Cody T.-H. Chiu
+ */
+
+#include <re.h>
+#include <baresip.h>
+#include <string.h>
+
+
+/**
+ * @defgroup pns pns
+ *
+ * Push Notification Service Module for SIP
+ */
+
+
+/** Defines PNS parameters  */
+struct pns {
+	char *pn_provider;	/**< Push provider */
+	char *pn_prid;		/**< Push resource ID */
+	char *pn_param;		/**< Push provider optional param */
+};
+
+
+static void destructor(void *data)
+{
+	struct pns *pns = data;
+
+	mem_deref(pns->pn_provider);
+	mem_deref(pns->pn_prid);
+	mem_deref(pns->pn_param);
+}
+
+
+/**
+ * Allocate a Push Notification Service
+ *
+ * @param pnsp     	Pointer to allocated Push Notification Service
+ * @param provider	Push provider
+ * @param prid    	Push resource ID
+ * @param param 	Push provider optional parameter
+ *
+ * @return 0 if success, otherwise errorcode
+ */
+static int pns_alloc(struct pns **pnsp, const struct pl *provider,
+		const struct pl *prid, const struct pl *param)
+{
+	struct pns *pns;
+	int err = 0;
+
+	if (!pl_isset(provider) || !pl_isset(prid)) {
+			warning("pns: provider and prid are required\n");
+		return EINVAL;
+	}
+
+	pns = mem_zalloc(sizeof(*pns), destructor);
```

Comment:
Please add a check if `pns` is allocated here, otherwise return `ENOMEM`;

Reply:
Done